File: src/ts/tests/assemblage-extra.test.ts
import { expect, test } from 'vitest';
import { Assemblage, Phrase, Strand } from '../model';
import { Instrument } from '@shared/enums';

// Strand error cases

test('Strand.addPhrase and removePhrase error handling', () => {
  const a = new Assemblage(Instrument.Sitar, 'A');
  a.addStrand('s');
  const strand = a.strands[0];
  const p = new Phrase();
  a.addPhrase(p); // register phrase in assemblage
  strand.addPhrase(p);
  expect(() => strand.addPhrase(p)).toThrow(Error);
  strand.removePhrase(p);
  expect(() => strand.removePhrase(p)).toThrow(Error);
});

// Assemblage.addStrand / addPhrase

test('Assemblage.addStrand and addPhrase duplicate and missing strand', () => {
  const a = new Assemblage(Instrument.Sitar, 'A');
  a.addStrand('dup');
  expect(() => a.addStrand('dup')).toThrow(Error);

  const p = new Phrase();
  a.addPhrase(p);
  expect(() => a.addPhrase(p)).toThrow(Error);
  expect(() => a.addPhrase(new Phrase(), 'missing')).toThrow(Error);
});

// Assemblage.removeStrand / removePhrase

test('Assemblage.removeStrand and removePhrase errors', () => {
  const a = new Assemblage(Instrument.Sitar, 'A');
  expect(() => a.removeStrand('bad')).toThrow(Error);
  const p = new Phrase();
  expect(() => a.removePhrase(p)).toThrow(Error);
});

// Assemblage.movePhraseToStrand branches

test('movePhraseToStrand removes from source when target missing', () => {
  const a = new Assemblage(Instrument.Sitar, 'A');
  a.addStrand('s1');
  const s1 = a.strands[0];
  const p = new Phrase();
  a.addPhrase(p, s1.id);
  a.movePhraseToStrand(p, 'none');
  expect(s1.phraseIDs).toEqual([]);
});

test('movePhraseToStrand errors if phrase not in assemblage', () => {
  const a = new Assemblage(Instrument.Sitar, 'A');
  a.addStrand('s1');
  const s1 = a.strands[0];
  const p = new Phrase();
  expect(() => a.movePhraseToStrand(p, s1.id)).toThrow(Error);
});

test('movePhraseToStrand moves phrase between strands', () => {
  const a = new Assemblage(Instrument.Sitar, 'A');
  a.addStrand('s1');
  a.addStrand('s2');
  const [s1, s2] = a.strands;
  const p = new Phrase();
  a.addPhrase(p, s1.id);
  a.movePhraseToStrand(p, s2.id);
  expect(s1.phraseIDs.includes(p.uniqueId)).toBe(false);
  expect(s2.phraseIDs).toEqual([p.uniqueId]);
});

// Assemblage.fromDescriptor validations

test('fromDescriptor throws on unknown phrase IDs', () => {
  const desc = {
    instrument: Instrument.Sitar,
    strands: [{ label: 's', phraseIDs: ['bad'], id: 'sid' }],
    name: 'A',
    id: 'id',
    loosePhraseIDs: []
  };
  expect(() => Assemblage.fromDescriptor(desc, [])).toThrow(Error);
});

test('fromDescriptor throws on unknown loose phrase IDs', () => {
  const desc = {
    instrument: Instrument.Sitar,
    strands: [],
    name: 'A',
    id: 'id',
    loosePhraseIDs: ['bad']
  };
  expect(() => Assemblage.fromDescriptor(desc, [])).toThrow(Error);
});

File: python/api/tests/assemblage_extra_test.py
import pytest

from python.api.classes.assemblage import Assemblage, Phrase, Instrument

# Strand error cases

def test_strand_add_remove_errors():
    a = Assemblage(Instrument.SITAR, 'A')
    a.add_strand('s')
    strand = a.strands[0]
    p = Phrase()
    a.add_phrase(p)  # register phrase in assemblage
    strand.add_phrase(p)
    with pytest.raises(ValueError):
        strand.add_phrase(p)
    strand.remove_phrase(p)
    with pytest.raises(ValueError):
        strand.remove_phrase(p)

# Assemblage.addStrand / addPhrase

def test_assemblage_add_duplicate_and_missing_strand():
    a = Assemblage(Instrument.SITAR, 'A')
    a.add_strand('dup')
    with pytest.raises(ValueError):
        a.add_strand('dup')

    p = Phrase()
    a.add_phrase(p)
    with pytest.raises(ValueError):
        a.add_phrase(p)
    with pytest.raises(ValueError):
        a.add_phrase(Phrase(), 'missing')

# Assemblage.removeStrand / removePhrase

def test_assemblage_remove_errors():
    a = Assemblage(Instrument.SITAR, 'A')
    with pytest.raises(ValueError):
        a.remove_strand('bad')
    p = Phrase()
    with pytest.raises(ValueError):
        a.remove_phrase(p)

# Assemblage.movePhraseToStrand branches

def test_move_phrase_to_strand_removes_from_source_when_target_missing():
    a = Assemblage(Instrument.SITAR, 'A')
    a.add_strand('s1')
    s1 = a.strands[0]
    p = Phrase()
    a.add_phrase(p, s1.id)
    a.move_phrase_to_strand(p, 'none')
    assert s1.phrase_ids == []


def test_move_phrase_to_strand_errors_if_phrase_not_in_assemblage():
    a = Assemblage(Instrument.SITAR, 'A')
    a.add_strand('s1')
    s1 = a.strands[0]
    p = Phrase()
    with pytest.raises(ValueError):
        a.move_phrase_to_strand(p, s1.id)


def test_move_phrase_to_strand_moves_phrase_between_strands():
    a = Assemblage(Instrument.SITAR, 'A')
    a.add_strand('s1')
    a.add_strand('s2')
    s1, s2 = a.strands
    p = Phrase()
    a.add_phrase(p, s1.id)
    a.move_phrase_to_strand(p, s2.id)
    assert p.unique_id not in s1.phrase_ids
    assert s2.phrase_ids == [p.unique_id]

# Assemblage.from_descriptor validations

def test_from_descriptor_throws_on_unknown_phrase_ids():
    desc = {
        'instrument': 'Sitar',
        'strands': [{'label': 's', 'phraseIDs': ['bad'], 'id': 'sid'}],
        'name': 'A',
        'id': 'id',
        'loosePhraseIDs': []
    }
    with pytest.raises(ValueError):
        Assemblage.from_descriptor(desc, [])


def test_from_descriptor_throws_on_unknown_loose_phrase_ids():
    desc = {
        'instrument': 'Sitar',
        'strands': [],
        'name': 'A',
        'id': 'id',
        'loosePhraseIDs': ['bad']
    }
    with pytest.raises(ValueError):
        Assemblage.from_descriptor(desc, [])
